data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/GenericEqualOperator.java

```diff
@@ -8,41 +8,41 @@
 import org.antlr.v4.runtime.RuleContext;
 import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser;
 
-import java.util.function.BiPredicate;
+import java.util.Map;
+import java.util.function.BiFunction;
 
 import static com.google.common.base.Preconditions.checkArgument;
 
-public class GenericEqualOperator implements Operator<Boolean> {
-    private final int symbol;
-    private final String displayName;
-    private final BiPredicate<Object, Object> operation;
+public class GenericEqualOperator extends BinaryOperator<Boolean> {
+    private final Map<Class<?>, Map<Class<?>, BiFunction<Object, Object, Boolean>>> equalStrategy;
 
-    public GenericEqualOperator(final int symbol, BiPredicate<Object, Object> operation) {
-        this.symbol = symbol;
-        displayName = DataPrepperExpressionParser.VOCABULARY.getDisplayName(symbol);
-        this.operation = operation;
+    public GenericEqualOperator(
+            final int symbol,
+            final int shouldEvaluateRuleIndex,
+            final Map<Class<?>, Map<Class<?>, BiFunction<Object, Object, Boolean>>> equalStrategy
```

Comment:
This seems a very rigid strategy. As the client of this code you must make a very complicated map.

Why not just use a `BiFunction<Object, Object, Boolean>` as the strategy? Then let the caller make a more complicated use of `Map` if necessary for that particular situation?

Reply:
It seems that all the `GenericEqualOperator` should care about is whether two objects are equal. Thus, the only strategy it needs is `BiPredicate<Object, Object>`. (I have another comment asking predicate vs. function below).

So this code should be able to run:

```
return operation.test(leftOperand, rightOperand);
```

The caller will determine how this `BiPredicate` works. In the current case, the caller does it by making the map. We could even have a common class to help with this.

```
public class ClassBasedBiFunction implements BiFunction<Object, Object> {
  private final Map<Class<?>, Map<Class<?>, BiPredicate<Object, Object>>> lhsToRhsToPredicate;
  private final BiPredicate<Object, Object> defaultOperation;
  public ClassBasedBiFunction(final Map<Class<?>, Map<Class<?>, BiPredicate<Object, Object>>> lhsToRhsToPredicate, BiPredicate<Object, Object> defaultOperation) {
  this. lhsToRhsToPredicate = lhsToRhsToPredicate;
  }

  public boolean test(Object lhs, Object rhs) {
    if (hasStrategyMatching(lhs, rhs)) {
             return equalStrategy.get(lhs.getClass())
                     .get(rhs.getClass())
                     .apply(lhs, rhs);
         }
         else {
             return defaultOperation.test(lhs, rhs);
         }
  }

  private boolean hasStrategyMatching(final Object leftOperand, final Object rightOperand) {
    // more or less the same
  }
}

Reply:
I see we are already using this pattern:

https://github.com/opensearch-project/data-prepper/blob/a37d557cf64740a98b53af2fd40a8324d8cec807/data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/NumericCompareOperator.java#L21-L22

So I'm fine with continuing it.

Reply:
Great suggestions on the BiPredicate, I was previously unfamiliar with that interface.